### ISSUE TITLE
Change property of a private unnamed type: from public to private

### ISSUE THREAD

**Issue by Its-Just-Nans** [obi1kenobi/cargo-semver-checks#1625]
### Is this about an existing lint, or proposing a new one?

It's about https://doc.rust-lang.org/beta/rustc/lints/listing/allowed-by-default.html?highlight=unnameable_types#unnameable-types

### Known issues that might be causing this

- [ ] Is the item that should have been flagged originating from another crate, or from a re-export of such an item? ([#638](https://github.com/obi1kenobi/cargo-semver-checks/issues/638))
- [ ] Is the issue related to a change in the type of a field, function parameter, generic, or trait bound? ([#149](https://github.com/obi1kenobi/cargo-semver-checks/issues/149))
- [ ] If you're proposing a new lint, is it already part of our lint wishlist? ([#5](https://github.com/obi1kenobi/cargo-semver-checks/issues/5))

### Steps to reproduce the bug with the above code

```rs
pub trait HasZipMetadata {
    // Required method
    fn get_metadata(&self) -> &ZipFileData;
}

mod priv {
      pub struct ZipFileData {
          pub value: u16
      }
}
```

if we change the `pub value: u16` to `value: u16`. cargo-semver-checks is okay with that. but this is a breaking change

Related to https://users.rust-lang.org/t/warning-when-building-public-trait-returning-private-struct/139526

### Actual Behaviour

Summary no semver update required

### Expected Behaviour

Breaking change

### Generated System Information

System information:
-------------------
#### Software version

cargo-semver-checks 0.47.0

#### Operating system

- OS: Linux (Ubuntu 24.04)
- Kernel: 6.17.0-20-generic

#### Command-line

```bash
/home/n4n5/.cargo/bin/cargo-semver-checks semver-checks --bugreport 
```

#### cargo version

```
> cargo -V 
cargo 1.94.0 (85eff7c80 2026-01-15)
```

#### Compile time information

- Profile: release
- Target triple: x86_64-unknown-linux-gnu
- Family: unix
- OS: linux
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: fxsr,sse,sse2
- Host: x86_64-unknown-linux-gnu

Cargo build configuration:
--------------------------
[build]
jobs = 12

### Additional Context

Related to https://users.rust-lang.org/t/warning-when-building-public-trait-returning-private-struct/139526

**Comment by Its-Just-Nans** [comment]
I think also for adding `#[non_exhaustive]` to a private unnamed type

**Comment by obi1kenobi** [comment]
Yes, good issue — thank you! You are correct: unnamable types have most of the same flavors of breakage that regular, nameable an importable types have, and we don't catch those cases right now.

One of the challenges is that the rules here aren't just "types returned by pub functions of pub types can result in SemVer breakage" but also that this applies recursively. We need to cover the transitive closure here. For example, below `Nested` is returned by `ZipFileData::get_nested()` which is itself not nameable directly, but is reachable via `HasZipMetadata::get_metadata()`:
```
pub trait HasZipMetadata {
    // Required method
    fn get_metadata(&self) -> &ZipFileData;
}

mod priv {
    pub struct ZipFileData {
        nested: Nested
    }

    impl ZipFileData {
        pub fn get_nested(&self) -> &Nested {
            &self.nested
        }
    }

    pub struct Nested {
       // Removing `pub` here is also breaking!
       pub value: u64
    }
}
```

I could probably carve out some good first issues in this area, if you or someone else might be interested in contributing to help close this gap!

**Comment by 2miwon** [comment]
Hi @obi1kenobi,

I am still relatively new to this specific area, but I would be truly grateful if I could help with this problem

I'm planning to start exploring the codebase locally to get a feel for how we can handle this recursively. Please let me know if you end up splitting this into smaller issues, or if you have any quick pointers before I start

**Comment by 2miwon** [comment]
I’d like to confirm the intended scope:

---

### Proposed Scope

- **Field visibility changes**: `pub` fields of unnameable `pub` structs becoming private.
- **Field deletion/rename**: Any change that makes a previously accessible field missing under its prior name.

This applies recursively — structs reachable only through another unnameable struct's method should also be covered.

### Open Question: Tuple Structs

Should this lint also cover unnameable tuple structs 
(e.g., `pub struct Data(pub u16, pub u32)`) where fields are accessed by index?

Does this look right?

**Comment by obi1kenobi** [comment]
Hi @2miwon, thanks for your enthusiasm!

I would strongly recommend *not* choosing this issue as one of your first contributions to the project. It is not a quick and easy fix, and requires a bunch of background context you probably don't have yet.

It is probably very tempting to stick it into an LLM and open a PR with the output. I *strongly* encourage you to not do that — the outcome will not be positive and will waste both of our time.

To meaningfully contribute to `cargo-semver-checks`, please consider:
- putting the LLM aside and digging into the code yourself
- writing some lints *by hand*
- getting some simpler PRs merged, one at a time, until you get the hang of contributing

If you do the above, I'm happy to mentor you. We can set solving this challenging issue as the target goal of your contributions. That'd be awesome! But I ask that you not just feed this issue and our conversations into an LLM and then make me review the output — if that's the approach, it'd be faster and less work if I just typed things into the LLM and did the work myself.
